### PR TITLE
docs: add CMM examples

### PR DESCRIPTION
**Pull request by mattsb42-aws**
*Issue #, if available:*
#156 

*Description of changes:*
Adds examples for using and creating CMMs.

As always, everyone is welcome to review anything, but what I'm specifically looking for from each reviewer:
* @juneb : Descriptions of each example
* @WesleyRosenblum and @acioc : Usefulness and scope of examples

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

### CODE REVIEW

**Comment by mattsb42-aws** [comment]
Note for the record: all failing checks are expected. They are all from the test vector handler workflow. The tests from that workflow actually should not be running on pull request events, but I'll figure that out in a different PR.

File: examples/src/crypto_materials_manager/explicit_default.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+"""
+The default cryptographic materials manager (CMM)
+is a CMM that uses a keyring or master key provider
+to manage the data key materials.
+When using the encrypt and decrypt APIs, as well as the caching CMM,
+you can supply either a CMM, a keyring, or a master key provider.
+If you supply a keyring or master key provider
+those components will automatically wrap it in a default CMM.
+
+This example shows how you could do this manually if you wanted to.
```
Comment (WesleyRosenblum) [inline examples/src/crypto_materials_manager/explicit_default.py:12]: This example I'm a little unsure is necessary. It is useful in terms of educating the user about the default CMM, but practically, without any additional functionality, it is pointless. Would it be enough to include the description of the default CMM in one of the other examples?

File: examples/src/crypto_materials_manager/custom/requiring_encryption_context_fields.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+This is especially powerful when you use the AWS Encryption SDK with AWS KMS,
+but within the context of the AWS Encryption SDK alone
+you can use cryptographic materials managers to analyse the encryption context
+to provide logical controls and additional metadata.
+
+https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/concepts.html#encryption-context
+https://docs.aws.amazon.com/kms/latest/developerguide/concepts.html#encrypt_context
+
+This example shows how to create a custom cryptographic materials manager (CMM)
+that requires a particular field in any encryption context.
```
[the remark this answers is not in the document]
Comment (mattsb42-aws) [reply examples/src/crypto_materials_manager/custom/requiring_encryption_context_fields.py:15]:
```suggestion
that requires a particular field in the encryption context.
```

File: examples/src/crypto_materials_manager/custom/algorithm_suite_enforcement.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        """Set up the inner cryptographic materials manager using the provided keyring.
+
+        :param Keyring keyring: Keyring to use in the inner cryptographic materials manager
+        """
+        self._cmm = DefaultCryptoMaterialsManager(keyring=keyring)
+
+    def get_encryption_materials(self, request):
+        # type: (EncryptionMaterialsRequest) -> EncryptionMaterials
+        """Block any requests that include an algorithm suite without a KDF."""
+        if request.algorithm is not None and request.algorithm.kdf is KDFSuite.NONE:
```
[the remark this answers is not in the document]
Comment (mattsb42-aws) [reply examples/src/crypto_materials_manager/custom/algorithm_suite_enforcement.py:47]: Maybe. If we want to go that route I want to re-frame what the example is about: changing it from specifically barring non-KDF suites to instead making sure that you can cache the request.

Another option would be changing the requirement to be one of our two recommended (default or default - signing).

[the remark this answers is not in the document]
Comment (WesleyRosenblum) [reply examples/src/crypto_materials_manager/custom/algorithm_suite_enforcement.py:47]: yeah, I'd say let's do the second option. The example doesn't have to do anything particularly interesting or innovative to serve its purpose, that would show a basic validation so the user can get a sense of whats possible.

File: examples/src/crypto_materials_manager/custom/requiring_encryption_context_fields.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
+# Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+"""
+Encryption context is a powerful tool for access and audit controls
+because it lets you tie *non-secret* metadata about a plaintext value to the encrypted message.
+This is especially powerful when you use the AWS Encryption SDK with AWS KMS,
+but within the context of the AWS Encryption SDK alone
```
[the remark this answers is not in the document]
Comment (mattsb42-aws) [reply examples/src/crypto_materials_manager/custom/requiring_encryption_context_fields.py:7]: I'd like to include some nod to KMS, to point people to "you can do X with this but Y if you're using KMS". I'll look at reworking this.

File: examples/src/crypto_materials_manager/explicit_default.py (hunk 1 above)
Comment (mattsb42-aws) [reply examples/src/crypto_materials_manager/explicit_default.py:12]: Maybe. What I was going for is demonstrating how to use the default CMM if you need it, such as if you're making a custom CMM that you configure using keyrings. More explanation in the other examples might be sufficient, yeah.

File: examples/src/crypto_materials_manager/custom/algorithm_suite_enforcement.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        "is not": "secret",
+        "but adds": "useful metadata",
+        "that can help you": "be confident that",
+        "the data you are handling": "is what you think it is",
+    }
+
+    # Create the keyring that determines how your data keys are protected.
+    keyring = KmsKeyring(generator_key_id=aws_kms_cmk)
+
+    # Create the filtering cryptographic materials manager using your keyring.
```
[the remark this answers is not in the document]
Comment (mattsb42-aws) [reply examples/src/crypto_materials_manager/custom/algorithm_suite_enforcement.py:82]: This was just shorthand I was using, but you're right; I'll just replace these with references back to the actual custom CMM.

File: examples/src/crypto_materials_manager/custom/algorithm_suite_enforcement.py [hunk 6]
```diff
@@ -47,9 +62,9 @@ def __init__(self, keyring):
 
     def get_encryption_materials(self, request):
         # type: (EncryptionMaterialsRequest) -> EncryptionMaterials
-        """Block any requests that include an algorithm suite without a KDF."""
-        if request.algorithm is not None and request.algorithm.kdf is KDFSuite.NONE:
-            raise UnsupportedAlgorithmSuite("Non-KDF algorithm suites are not allowed!")
+        """Block any requests that include an approved algorithm suite."""
```
Comment (WesleyRosenblum) [inline examples/src/crypto_materials_manager/custom/algorithm_suite_enforcement.py:65]:
```suggestion
        """Block any requests that include an unapproved algorithm suite."""
```

File: examples/src/crypto_materials_manager/custom/algorithm_suite_enforcement.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+If your readers and writers have the same permissions,
+you might want to omit the message signature for faster operation.
+For more information about choosing a signed or unsigned algorithm suite,
+see the AWS Encryption SDK developer guide:
+
+https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/supported-algorithms.html#other-algorithms
+
+This example shows how you can make a custom cryptographic materials manager (CMM)
+that only allows encrypt requests that either specify one of these two algorithm suites
+or do not specify an algorithm suite, in which case the default CMM uses the default algorithm suite,
```
Comment (WesleyRosenblum) [inline examples/src/crypto_materials_manager/custom/algorithm_suite_enforcement.py:23]:
```suggestion
or do not specify an algorithm suite, in which case the default CMM uses the default algorithm suite.
```

**Pull request merged**